File: templates/typescript-fastify/src/shared/types/result.ts
/**
 * BRIK v5 Result Type - Functional error handling
 */

export abstract class Result<T, E extends Error> {
  abstract isOk(): this is Ok<T, E>;
  abstract isErr(): this is Err<T, E>;
  abstract unwrap(): T;
  abstract unwrapOr(defaultValue: T): T;
  abstract unwrapErr(): E;
  
  map<U>(fn: (value: T) => U): Result<U, E> {
    if (this.isOk()) {
      return Ok.of(fn(this.value));
    }
    return Err.of(this.error);
  }
  
  mapErr<F extends Error>(fn: (error: E) => F): Result<T, F> {
    if (this.isErr()) {
      return Err.of(fn(this.error));
    }
    return Ok.of(this.value);
  }
  
  and<U>(other: Result<U, E>): Result<U, E> {
    if (this.isOk()) {
      return other;
    }
    return Err.of(this.error);
  }
  
  or(other: Result<T, E>): Result<T, E> {
    if (this.isOk()) {
      return this;
    }
    return other;
  }
  
  static ok<T, E extends Error>(value: T): Result<T, E> {
    return Ok.of(value);
  }
  
  static err<T, E extends Error>(error: E): Result<T, E> {
    return Err.of(error);
  }
}

export class Ok<T, E extends Error> extends Result<T, E> {
  constructor(public readonly value: T) {
    super();
  }
  
  isOk(): this is Ok<T, E> {
    return true;
  }
  
  isErr(): this is Err<T, E> {
    return false;
  }
  
  unwrap(): T {
    return this.value;
  }
  
  unwrapOr(_defaultValue: T): T {
    return this.value;
  }
  
  unwrapErr(): E {
    throw new Error('Called unwrapErr on Ok result');
  }
  
  static of<T, E extends Error>(value: T): Ok<T, E> {
    return new Ok(value);
  }
}

export class Err<T, E extends Error> extends Result<T, E> {
  constructor(public readonly error: E) {
    super();
  }
  
  isOk(): this is Ok<T, E> {
    return false;
  }
  
  isErr(): this is Err<T, E> {
    return true;
  }
  
  unwrap(): T {
    throw this.error;
  }
  
  unwrapOr(defaultValue: T): T {
    return defaultValue;
  }
  
  unwrapErr(): E {
    return this.error;
  }
  
  static of<T, E extends Error>(error: E): Err<T, E> {
    return new Err(error);
  }
}

// Utility functions
export const ok = Result.ok;
export const err = Result.err;

File: templates/rust-axum/src/shared/types/result.rs
//! BRIK v5 Result Type - Functional error handling for Rust

use std::fmt::Debug;

/// A type that represents either success (`Ok`) or failure (`Err`).
/// Similar to std::result::Result but with additional helper methods for BRIK patterns.
pub type BrikResult<T, E> = std::result::Result<T, E>;

/// Extension trait for Result to add BRIK-specific functionality
pub trait ResultExt<T, E> {
    /// Maps a Result<T, E> to Result<U, E> by applying a function to a contained Ok value
    fn map_ok<U, F>(self, f: F) -> BrikResult<U, E>
    where
        F: FnOnce(T) -> U;

    /// Maps a Result<T, E> to Result<T, F> by applying a function to a contained Err value
    fn map_err_to<F, G>(self, f: G) -> BrikResult<T, F>
    where
        G: FnOnce(E) -> F;

    /// Returns the contained Ok value or a provided default
    fn unwrap_or_else<F>(self, f: F) -> T
    where
        F: FnOnce(E) -> T;

    /// Returns Ok if both self and other are Ok, otherwise returns the first Err
    fn and_then_combine<U, F>(self, other: BrikResult<U, E>, f: F) -> BrikResult<(T, U), E>
    where
        F: FnOnce() -> BrikResult<(T, U), E>,
        E: Clone;
}

impl<T, E> ResultExt<T, E> for BrikResult<T, E> {
    fn map_ok<U, F>(self, f: F) -> BrikResult<U, E>
    where
        F: FnOnce(T) -> U,
    {
        self.map(f)
    }

    fn map_err_to<F, G>(self, f: G) -> BrikResult<T, F>
    where
        G: FnOnce(E) -> F,
    {
        self.map_err(f)
    }

    fn unwrap_or_else<F>(self, f: F) -> T
    where
        F: FnOnce(E) -> T,
    {
        match self {
            Ok(value) => value,
            Err(error) => f(error),
        }
    }

    fn and_then_combine<U, F>(self, other: BrikResult<U, E>, _f: F) -> BrikResult<(T, U), E>
    where
        F: FnOnce() -> BrikResult<(T, U), E>,
        E: Clone,
    {
        match (self, other) {
            (Ok(a), Ok(b)) => Ok((a, b)),
            (Err(e), _) => Err(e),
            (_, Err(e)) => Err(e),
        }
    }
}

/// Convenience function to create an Ok result
pub fn ok<T, E>(value: T) -> BrikResult<T, E> {
    Ok(value)
}

/// Convenience function to create an Err result
pub fn err<T, E>(error: E) -> BrikResult<T, E> {
    Err(error)
}

/// Macro to convert multiple Results into a single Result containing a tuple
/// Usage: combine_results!(result1, result2, result3)
#[macro_export]
macro_rules! combine_results {
    ($first:expr) => {
        $first.map(|v| (v,))
    };
    ($first:expr, $($rest:expr),+) => {
        {
            let first = $first?;
            let rest = combine_results!($($rest),+)?;
            Ok({
                let mut tuple = (first,);
                // This is a bit hacky but works for combining tuples
                #[allow(unused_assignments)]
                {
                    tuple = (tuple.0, rest.0);
                    $(
                        // Add more tuple elements as needed
                    )+
                }
                tuple
            })
        }
    };
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_result_ext_map_ok() {
        let result: BrikResult<i32, &str> = Ok(5);
        let mapped = result.map_ok(|x| x * 2);
        assert_eq!(mapped, Ok(10));
    }

    #[test]
    fn test_result_ext_map_err() {
        let result: BrikResult<i32, &str> = Err("error");
        let mapped = result.map_err_to(|e| format!("Mapped: {}", e));
        assert_eq!(mapped, Err("Mapped: error".to_string()));
    }

    #[test]
    fn test_ok_err_helpers() {
        let success: BrikResult<i32, String> = ok(42);
        let failure: BrikResult<i32, String> = err("failed".to_string());

        assert_eq!(success, Ok(42));
        assert_eq!(failure, Err("failed".to_string()));
    }
}